OBSDOCS-270: Reference doc for remote write configuration isn't exhau… (#80449)
openshift/openshift-docs: modules/monitoring-example-remote-write-queue-configuration.adoc

:prometheus-instance: prometheusK8s
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
:configmap-name: user-workload-monitoring-config
:namespace-name: openshift-user-workload-monitoring
:prometheus-instance: prometheus
endif::openshift-dedicated,openshift-rosa[]

You can use the `queueConfig` object for remote write to tune the remote write queue parameters. The following is the example configuration of the queue parameters for 
ifndef::openshift-dedicated,openshift-rosa[]
default platform monitoring
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
monitoring for user-defined projects
endif::openshift-dedicated,openshift-rosa[]
in the `{namespace-name}` namespace.

.Example configuration of remote write parameters
[source,yaml,subs="attributes+"]

----
apiVersion: v1
kind: ConfigMap
metadata:
  name: {configmap-name}
  namespace: {namespace-name}
data:
  config.yaml: |
    {prometheus-instance}:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com" 
        <endpoint_authentication_credentials>
        queueConfig:
          #example parameters #<1>
----

> <1> TBD

Reviewer: * capacity: the number of samples to buffer per shard before dropping them.
* minShards: the minimum number of shards, i.e. amount of concurrency.
* maxShards: the maximum number of shards, i.e. amount of concurrency.
* maxSamplesPerSend: the maximum number of samples per send.
* batchSendDeadline: the maximum time a sample will wait in buffer.
* minBackoff: the initial retry delay. It gets doubled for every retry.
* maxBackoff: the maximum retry delay.
* retryOnRateLimit: whether to retry upon receiving a 429 status code from the remote-write storage.
* sampleAgeLimit: drops samples older than the limit.

// * observability/monitoring/configuring-the-monitoring-stack.adoc

:_mod-docs-content-type: REFERENCE
[id="example-remote-write-queue-configuration_{context}"]
= Example remote write queue configuration

// Set attributes to distinguish between cluster monitoring examples and user workload monitoring examples.
ifndef::openshift-dedicated,openshift-rosa[]
:configmap-name: cluster-monitoring-config
:namespace-name: openshift-monitoring
:prometheus-instance: prometheusK8s
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
:configmap-name: user-workload-monitoring-config
:namespace-name: openshift-user-workload-monitoring
:prometheus-instance: prometheus
endif::openshift-dedicated,openshift-rosa[]

You can use the `queueConfig` object for remote write to tune the remote write queue parameters. The following example shows the queue parameters with their default values for 
ifndef::openshift-dedicated,openshift-rosa[]
default platform monitoring
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
monitoring for user-defined projects
endif::openshift-dedicated,openshift-rosa[]
in the `{namespace-name}` namespace.

> .Example configuration of remote write parameters with default values

Reviewer 2: Should this be a heading as you did in the [other module](https://github.com/openshift/openshift-docs/pull/80449/files#r1768610889)?

Author: Thank you for asking! This is a simple example so there is no need to have it as title. The titled ones are based on other situation. We use it in monitoring docs for when we wanted to present samples that contain 2 and more files (because it would be hard to distinguish without it.

[source,yaml,subs="attributes+"]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: {configmap-name}
  namespace: {namespace-name}
data:
  config.yaml: |
    {prometheus-instance}:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com" 
        <endpoint_authentication_credentials>
        queueConfig:
          capacity: 10000 #<1>
          minShards: 1 #<2>
          maxShards: 50 #<3>
          maxSamplesPerSend: 2000 #<4>
          batchSendDeadline: 5s #<5>
          minBackoff: 30ms #<6>
          maxBackoff: 5s #<7>
          retryOnRateLimit: false #<8>
          sampleAgeLimit: 0s #<9>
----
<1> The number of samples to buffer per shard before they are dropped from the queue.
<2> The minimum number of shards.
<3> The maximum number of shards.
<4> The maximum number of samples per send.
<5> The maximum time for a sample to wait in buffer.

> <6> The initial time to wait before retrying a failed request. It gets doubled for every retry up to the `maxbackoff` time.

Reviewer 2: Slightly ambiguous "it".
```suggestion
<6> The initial time to wait before retrying a failed request. The time gets doubled for every retry up to the `maxbackoff` time.
```

> <6> The initial time to wait before retrying a failed request. It gets doubled for every retry up to the `max_backoff` time.

Reviewer 3: ```suggestion
<6> The initial time to wait before retrying a failed request. It gets doubled for every retry up to the `maxBackoff` time.
```
is it fine?

<7> The maximum time to wait before retrying a failed request.

> <8> Set this parameter to `true` if you want to retry a request after receiving a 429 status code from the remote write storage.

Reviewer 3: do we need to mention it's an *experimental feature*?
```
# oc explain prometheus.spec.remoteWrite.queueConfig.retryOnRateLimit
GROUP:      monitoring.coreos.com
KIND:       Prometheus
VERSION:    v1

FIELD: retryOnRateLimit <boolean>

DESCRIPTION:
    Retry upon receiving a 429 status code from the remote-write storage.
    
    
    This is an *experimental feature*, it may change in any upcoming release
    in a breaking way.
```

Author: @simonpasquier would we concider this a TP feature? Otherwise, I would probably avoid including this one.

Reviewer: the description comes from the upstream CRD which itself mimics the upstream Prometheus documentation. I have no good answer for this but in practice:
* It's unlikely that the feature goes away.
* Even if it was, we would continue supporting it in the CMO configmap.

> <9> The samples that are older than the `sampleAgeLimit` limit are dropped from the queue. When the value is set to `0s` the parameter is ignored.

Author: I made the last sentence up :smile: is my guesswork here correct?

Author: Because I noticed that the default is zero, which might be confusing if we do not explain.

Reviewer: you're correct! Maybe we should say "when set to 0s or undefined"?

> <9> The samples that are older than the `sampleAgeLimit` limit are dropped from the queue. When the value is undefined or set to `0s`, the parameter is ignored.

Reviewer 2: ```suggestion
<9> The samples that are older than the `sampleAgeLimit` limit are dropped from the queue. If the value is undefined or set to `0s`, the parameter is ignored.
```
